### PR TITLE
feat: Add Frequent Users feature in [AtlasProxy]

### PR DESCRIPTION
**Pull request by mgorsk1**
### Summary of Changes

This change enables Frequent Users feature in Table Detail View and My Profile section.

### Tests

_What tests did you add or modify and why? If no tests were added or modified, explain why. Remove this line_

### Documentation

_What documentation did you add or modify and why? Add any relevant links then remove this line_

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes. 
- [x] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does
- [x] PR passes `make test`

### CODE REVIEW

**Comment by feng-tao** [comment]
cc @verdan

**Comment by feng-tao** [comment]
hey. @verdan , could you help to take a look?

File: metadata_service/proxy/atlas_proxy.py [hunk 1]
```diff
@@ -652,6 +689,66 @@ def _parse_date(self, date: int) -> Optional[int]:
...
+        for record in search_results.entities:
+            readers.append(record.guid)
+
+        results = []
+
+        if len(readers) > 0:
+            full_entities = AtlasProxy._extract_entities(self._driver.entity_bulk(guid=readers,
+                                                                                  ignoreRelationships=False))
+
+            for e in full_entities:
```
Comment (verdan) [inline metadata_service/proxy/atlas_proxy.py:744]: `e` -> `entity`
`r` -> `reader`

File: metadata_service/proxy/atlas_proxy.py [hunk 2]
```diff
@@ -652,6 +689,66 @@ def _parse_date(self, date: int) -> Optional[int]:
...
+                r = Reader(user=User(email=e.relationshipAttributes['user']['displayText']),
```
Comment (verdan) [inline metadata_service/proxy/atlas_proxy.py:745]: Please fill in the `user_id` field instead. We'll be making the email as optional and user_id as a required attribute. and also, the whole communication of frontend goes through `user_id`.

File: metadata_service/proxy/atlas_proxy.py [hunk 3]
```diff
@@ -652,6 +689,66 @@ def _parse_date(self, date: int) -> Optional[int]:
...
+            'attributes': [self.READER_COUNT_KEY, self.QN_KEY],
+            'sortBy': self.READER_COUNT_KEY,
+            'sortOrder': 'DESCENDING'
+        }
+
+        search_results = self._driver.search_basic.create(data=params, ignoreRelationships=False)
+
+        readers = []
+
+        for record in search_results.entities:
```
Comment (verdan) [inline metadata_service/proxy/atlas_proxy.py:735]: I think the user ID will also be available in `record` here. under related attributes, as the qualified name of User. you don't need to make extra call based on reader IDs, just for the user id. Can you please check ?

File: metadata_service/proxy/atlas_proxy.py [hunk 4]
```diff
@@ -596,8 +599,42 @@ def get_table_by_user_relation(self, *, user_email: str, relation_type: UserReso
 
         return {'table': results}
 
-    def get_frequently_used_tables(self, *, user_email: str) -> Dict[str, Any]:
-        pass
+    def get_frequently_used_tables(self, *, user_email: str) -> Dict[str, List[PopularTable]]:
+        user = self._driver.entity_unique_attribute(self.USER_TYPE, qualifiedName=user_email).entity
+
+        readers_guids = []
+        for r in user['relationshipAttributes'].get('entityReads'):
```
Comment (verdan) [inline metadata_service/proxy/atlas_proxy.py:606]: `r` -> `user_reads`

File: metadata_service/proxy/atlas_proxy.py [hunk 5]
```diff
@@ -596,8 +599,42 @@ def get_table_by_user_relation(self, *, user_email: str, relation_type: UserReso
...
+                readers_guids.append(r['guid'])
+
+        readers = self._extract_entities(self._driver.entity_bulk(guid=readers_guids, ignoreRelationships=True))
+
+        _results = {}
+        for reader in readers:
+            entity_uri = reader.attributes.get(self.ENTITY_URI_KEY)
+            count = reader.attributes.get('count')
+
+            if count > 0:
```
Comment (verdan) [inline metadata_service/proxy/atlas_proxy.py:620]: same as `if count:`

File: metadata_service/proxy/atlas_proxy.py [hunk 6]
```diff
@@ -596,8 +599,42 @@ def get_table_by_user_relation(self, *, user_email: str, relation_type: UserReso
...
+                details = self._extract_info_from_uri(table_uri=entity_uri)
+
+                _results[count] = dict(cluster=details.get('cluster'),
+                                       name=details.get('name'),
+                                       schema=details.get('db'),
+                                       database=details.get('entity'))
+
+        sorted_counts = list(_results.keys())
```
Comment (verdan) [inline metadata_service/proxy/atlas_proxy.py:628]: you mean `sorted(_results.keys())` to sort it. 
You can also use OrderedDict.
```
import collections
for key, value in collections.OrderedDict(sorted(_results.items())):
```

File: metadata_service/proxy/atlas_proxy.py [hunk 7]
```diff
@@ -652,6 +689,66 @@ def _parse_date(self, date: int) -> Optional[int]:
         except Exception:
             return None
 
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:692]: I used the same exact method that we have in amundsensearch (it was introduced by Nanne over a year ago). I could either make it classmethod or move to amundsencommon.

File: metadata_service/proxy/atlas_proxy.py (hunk 1 above)
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:744]: true, fixed

File: metadata_service/proxy/atlas_proxy.py (hunk 3 above)
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:735]: I think basic search doesn't return relationship attributes. I checked it and the attribute `relationshipAttributes` is always None

File: metadata_service/proxy/atlas_proxy.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (verdan) [reply metadata_service/proxy/atlas_proxy.py:692]: You can also open a PR in `pyatlasclient` for all the common functions you think could be used by others. 
I am the author of that repo, and can merge and deploy that pretty quickly! 😉

File: metadata_service/proxy/atlas_proxy.py (hunk 4 above)
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:606]: done

File: metadata_service/proxy/atlas_proxy.py (hunk 5 above)
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:620]: changed

File: metadata_service/proxy/atlas_proxy.py (hunk 6 above)
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:628]: agree, should be sorted. I am not sure though if OrderedDict is of any use here - it should be used to `remember the order entries were added` and that's not of any use here. we need to sort the keys by value not order of inserting

File: metadata_service/proxy/atlas_proxy.py (hunk 2 above)
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:745]: that was first implementation but turns out reader schema expects email, not user_id:

` "Encountered exception: {'table_readers': {0: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 1: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 2: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 3: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 4: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 5: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 6: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 7: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}}}"`

File: metadata_service/proxy/atlas_proxy.py [hunk 8]
```diff
@@ -41,12 +41,14 @@ class AtlasProxy(BaseProxy):
...
     BOOKMARK_TYPE = 'Bookmark'
     USER_TYPE = 'User'
+    READER_TYPE = 'Reader'
     QN_KEY = 'qualifiedName'
     BOOKMARK_ACTIVE_KEY = 'active'
     GUID_KEY = 'guid'
     ATTRS_KEY = 'attributes'
     REL_ATTRS_KEY = 'relationshipAttributes'
     ENTITY_URI_KEY = 'entityUri'
+    READER_COUNT_KEY = 'count'
```
[the remark this answers is not in the document]
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:51]: but if the name of the field changes you don't need to look for it in whole class. maybe we could use separate class like AtlasProxyConstants and move it there ?

**Comment by mgorsk1** [comment]
thanks for the review @verdan , addressed comments and provided fixes, good for the second look :muscle:

File: metadata_service/proxy/atlas_proxy.py (hunk 2 above)
Comment (verdan) [reply metadata_service/proxy/atlas_proxy.py]: oh yes, so you could fill in user_id and display_name/email with the CK. Will have a custom configurable function to get the user details wherever needed in the future. Main point is to fill in the user_id here.

Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py]: done

File: metadata_service/proxy/atlas_proxy.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (verdan) [reply metadata_service/proxy/atlas_proxy.py:51]: yes, but my point is that these are the things that usually don't change, and is custom defined in our entity definition so we also don't need to take care of the Atlas changes.

File: metadata_service/proxy/atlas_proxy.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:691]: ok, opened a PR https://github.com/verdan/pyatlasclient/pull/2

File: metadata_service/proxy/atlas_proxy.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py]: consider it done

**Comment by mgorsk1** [comment]
also waiting for https://github.com/lyft/amundsenmetadatalibrary/pull/146 to go in and will rebase, seems useful from our perspective

File: metadata_service/proxy/atlas_proxy.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py]: done

File: metadata_service/proxy/atlas_proxy.py [hunk 9]
```diff
@@ -652,6 +689,53 @@ def _parse_date(self, date: int) -> Optional[int]:
...
+        search_results = self._driver.search_basic.create(data=params, ignoreRelationships=False)
+
+        readers = []
+
+        for record in search_results.entities:
+            readers.append(record.guid)
+
+        results = []
+
+        if len(readers) > 0:
```
Comment (verdan) [inline metadata_service/proxy/atlas_proxy.py:727]: `if readers: `

File: metadata_service/proxy/atlas_proxy.py [hunk 10]
```diff
@@ -652,6 +689,53 @@ def _parse_date(self, date: int) -> Optional[int]:
...
+            full_entities = extract_entities(self._driver.entity_bulk(guid=readers, ignoreRelationships=False))
```
Comment (verdan) [inline metadata_service/proxy/atlas_proxy.py:728]: `full_entities` -> `read_entities`

```python
reader_entities = extract_entities(self._driver.entity_bulk(guid=readers, ignoreRelationships=False))
for entity in reader_entities:
```

File: metadata_service/proxy/atlas_proxy.py (hunk 9 above)
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:727]: fixed

File: metadata_service/proxy/atlas_proxy.py [hunk 11]
```diff
@@ -652,6 +689,53 @@ def _parse_date(self, date: int) -> Optional[int]:
         except Exception:
             return None
 
+    def _get_readers(self, qualified_name: str, top: Optional[int] = None) -> List[Reader]:
```
Comment (feng-tao) [inline metadata_service/proxy/atlas_proxy.py:692]: small nit: 
top: Optional[int] = 15? 
so you don't need to do top or 15 in L696

File: metadata_service/proxy/atlas_proxy.py [hunk 12]
```diff
@@ -596,8 +599,42 @@ def get_table_by_user_relation(self, *, user_email: str, relation_type: UserReso
 
         return {'table': results}
 
-    def get_frequently_used_tables(self, *, user_email: str) -> Dict[str, Any]:
-        pass
+    def get_frequently_used_tables(self, *, user_email: str) -> Dict[str, List[PopularTable]]:
```
Comment (feng-tao) [inline metadata_service/proxy/atlas_proxy.py:602]: is it easy to add a test for this function?

Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py:602]: sure, added tests

File: metadata_service/proxy/atlas_proxy.py (hunk 11 above)
Comment (mgorsk1) [reply metadata_service/proxy/atlas_proxy.py]: fixed

**Pull request merged**